airbyte-cdk/bulk/core/load/src/main/kotlin/io/airbyte/cdk/write/TaskRunner.kt

```diff
@@ -0,0 +1,121 @@
+package io.airbyte.cdk.write
+
+import io.github.oshai.kotlinlogging.KotlinLogging
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicInteger
+import kotlinx.coroutines.channels.*
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.yield
+
+/**
+ * Runs a DestinationTask state machine, either from an arbitrary initial task
+ * or by wrapping a StandardDestination in the default state machine entry point.
+ */
+class TaskRunner(
+    private val initialTask: AbstractDestinationTask<*, *>
+) {
+    constructor(destination: StandardDestination): this(
+        DefaultSetup(destination)
+    )
+
+    private val log = KotlinLogging.logger {}
+
+    // TODO: From configuration
+    companion object {
+        const val MAX_TASKS = 10
+        const val WAIT_TIME_MS = 500L
+    }
+
+    private val catalog = DummyCatalog() // TODO: from real catalog
+    private val streamComplete = mutableMapOf<Stream, Boolean>()
+
+    private val queue = Channel<AbstractDestinationTask<*, *>>(Channel.UNLIMITED)
+    private val nTasks = AtomicInteger(0)
+    private val nTasksPerType = mutableMapOf<String, AtomicInteger>()
+    private val nTasksPerTypePerStream = mutableMapOf<String, MutableMap<Stream, AtomicInteger>>()
+    private val seenUniqueIds = mutableSetOf<String>()
+    private val taskCounters = mutableMapOf<String, Int>()
```

Comment:
These map and set implementations are not thread-safe.

Reply:
they don't need to be, they're only accessed from the main thread

Reply:
I disagree. I see a call to `nTasksPerType.getOrPut` in `suspend fun run`. Once you're in `suspend`, you don't know which thread you're running on.

Reply:
Yes, you're right. I was confusing parallel-safe with thread-safe. 

Because suspend functions can switch threads, I should treat any two accesses on either side of a suspend point as if they were accesses from separate threads.

I'll attend to that throughout.